### PR TITLE
Bug 1745767: Fix a corner case in mergedMirrorSets

### PR DESCRIPTION
**Pull request by mtrmac**
**- What I did**
`RepositoryDigestMirrors` configuration with mirrors listing only the source (like in the added test case) was causing a panic; fix that.

**- How to verify it**
The added unit test.

**- Description for the changelog**
Fixed a panic in a corner case of ImageContentSourcePolicy mirror configuration.

### CODE REVIEW

**Comment by runcom** [comment]
@mtrmac you need a BZ targeting 4.2 in order to merge this for 4.2. Please create one and attach it to this PR.

**Comment by mtrmac** [comment]
> @mtrmac you need a BZ targeting 4.2 in order to merge this for 4.2. Please create one and attach it to this PR.

I’m fine with this only happening later; I think a stern “don’t do that” would be a reasonable response instead of rushing in a change into a frozen product.

(I’ll create a bug if you think this should get into 4.2, though.)

**Comment by mtrmac** [comment]
(Also, https://github.com/openshift/openshift-controller-manager/pull/19 will probably cause all of this code to be moved $somewhere soon.)

**Comment by mtrmac** [comment]
/hold

This still looks suboptimal for
```go
{Source: "source.example.net", Mirrors: []string{"source.example.net", "source.example.net", "source.example.net"}},
```
I’m not sure what the cleanest solution is, I’ll try to revisit this next week.

**Comment by runcom** [comment]
> I’m fine with this only happening later; I think a stern “don’t do that” would be a reasonable response instead of rushing in a change into a frozen product.

Uhm, I have little insight into this but a panic sounds reasonable enough to have a fix for 4.2 (I'm guessing but I don't have the full picture here). Product isn't frozen to bug fixes yet also.

Anyway, if you think this is severe enough and want to have this code landed feel free to create the BZ otherwise sure it can skip 4.2 (but I'm kind of feeling a panic should be fixed).

**Comment by runcom** [comment]
> @runcom Filed https://bugzilla.redhat.com/show_bug.cgi?id=1745767 , so that this gets in before moving the code into a separate subpackage (see #1087 / [openshift/openshift-controller-manager#19](https://github.com/openshift/openshift-controller-manager/pull/19) ). Please point me at process documentation if I should be doing more.

Thanks a lot @mtrmac - we should be all set here as well as mirrors are 4.2 so the BZ is ok and we can take this.

File: pkg/controller/container-runtime-config/helpers_test.go [hunk 1]
```diff
@@ -111,6 +137,17 @@ func TestMergedMirorSets(t *testing.T) {
 				{Source: "source.example.com", Mirrors: []string{"z1.example.com", "source.example.com", "y2.example.com", "x3.example.com"}},
 			},
 		},
+		{
+			// Pathological case of the above: _only_ the source included in mirrors, even perhaps several times.
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply pkg/controller/container-runtime-config/helpers_test.go]: http://catb.org/jargon/html/P/pathological.html all three senses.

I’ll no with any suggestion from native speakers, of course.

I can use “extreme” or “worst” instead if that helps, though that seems less expressive to me.

File: pkg/controller/container-runtime-config/helpers_test.go [hunk 2]
```diff
@@ -39,6 +39,32 @@ func TestScopeMatchesRegistry(t *testing.T) {
...
+				Source:  source,
+				Mirrors: tt.mirrors,
+			}
+			res := rdmContainsARealMirror(&set)
+			assert.Equal(t, tt.expected, res)
+		})
+	}
+}
+
 func TestMergedMirorSets(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply pkg/controller/container-runtime-config/helpers_test.go]: Thanks, fixed.

File: pkg/controller/container-runtime-config/helpers_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply pkg/controller/container-runtime-config/helpers_test.go]: > I’ll no with any suggestion from native speakers, of course.

I’ll **go** with …

[the remark this answers is not in the document]
Comment (mtrmac) [reply pkg/controller/container-runtime-config/helpers_test.go]: Updated to “worst case” per separate conversation.
